cmd/kola/kola.go

```diff
@@ -204,58 +208,88 @@ func writeProps() error {
 }
 
 func runList(cmd *cobra.Command, args []string) {
-	var w = tabwriter.NewWriter(os.Stdout, 0, 8, 0, '\t', 0)
-	var testlist []item
-
+	var testlist []*item
 	for name, test := range register.Tests {
-		testlist = append(testlist, item{
+		item := &item{
 			name,
 			test.Platforms,
 			test.ExcludePlatforms,
-			test.Architectures})
+			test.Architectures,
+			test.Distros,
+			test.ExcludeDistros}
+		item.updateValues()
+		testlist = append(testlist, item)
 	}
 
 	sort.Slice(testlist, func(i, j int) bool {
 		return testlist[i].Name < testlist[j].Name
 	})
 
-	fmt.Fprintln(w, "Test Name\tPlatforms\tArchitectures")
-	fmt.Fprintln(w, "\t")
-	for _, item := range testlist {
-		fmt.Fprintf(w, "%v\n", item)
+	if !listJSON {
+		var w = tabwriter.NewWriter(os.Stdout, 0, 8, 0, '\t', 0)
+
+		fmt.Fprintln(w, "Test Name\tPlatforms\tArchitectures\tDistributions")
+		fmt.Fprintln(w, "\t")
+		for _, item := range testlist {
+			fmt.Fprintf(w, "%v\n", item)
+		}
+		w.Flush()
+	} else {
+		out, err := json.MarshalIndent(testlist, "", "\t")
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "marshalling test list: %v\n", err)
+			os.Exit(1)
+		}
+		fmt.Println(string(out))
 	}
-	w.Flush()
 }
 
 type item struct {
 	Name             string
 	Platforms        []string
-	ExcludePlatforms []string
+	ExcludePlatforms []string `json:"-"`
 	Architectures    []string
+	Distros          []string
+	ExcludeDistros   []string `json:"-"`
 }
 
-func (i item) String() string {
-	if len(i.ExcludePlatforms) > 0 {
-		excludePlatforms := map[string]struct{}{}
-		for _, platform := range i.ExcludePlatforms {
-			excludePlatforms[platform] = struct{}{}
-		}
-		if len(i.Platforms) == 0 {
-			i.Platforms = kolaPlatforms
+func (i *item) updateValues() {
+	buildItems := func(include, exclude, all []string) []string {
+		var retItems []string
+		if len(exclude) > 0 {
+			excludeMap := map[string]struct{}{}
+			for _, item := range exclude {
+				excludeMap[item] = struct{}{}
+			}
+			if len(include) == 0 {
+				retItems = all
+			} else {
+				retItems = include
+			}
+			items := []string{}
+			for _, item := range retItems {
+				if _, ok := excludeMap[item]; !ok {
+					items = append(items, item)
+				}
+			}
+			retItems = items
+		} else {
+			retItems = include
 		}
-		platforms := []string{}
-		for _, platform := range i.Platforms {
-			if _, ok := excludePlatforms[platform]; !ok {
-				platforms = append(platforms, platform)
+		if len(retItems) == 0 {
```

Comment:
If `exclude` contains every element of `all`, this concludes that the test will always run, which doesn't match the behavior of `filterTests()`.

Reply:
Good catch, I'll add a case for ensuring that `exclude` isn't the same as `include`.

Reply:
`exclude` could also directly contain everything in `all`, inadvertently or otherwise.

Reply:
Should be addressed now, added a `sliceCompare` function and use it at the beginning of `buildItems` to compare if `exclude` & `all` are equal (returning an empty string list immediately if this is the case).